com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerVisualizer.cs

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.MixedReality.OpenXR;
+using Microsoft.MixedReality.Toolkit.Input.Simulation;
+using UnityEngine;
+using UnityEngine.InputSystem.Utilities;
+using UnityEngine.XR;
+using UnityEngine.XR.Interaction.Toolkit;
+using UnityInputSystem = UnityEngine.InputSystem;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Basic controller visualizer which draws a generic motion controller when one is detected
+    /// </summary>
+    [AddComponentMenu("MRTK/Input/Controller Visualizer")]
+    public class ControllerVisualizer : MonoBehaviour
+    {
+        [SerializeField]
+        [Tooltip("The XRNode on which this hand is located.")]
+        private XRNode handNode = XRNode.LeftHand;
+
+        /// <summary> The XRNode on which this hand is located. </summary>
+        public XRNode HandNode { get => handNode; set => handNode = value; }
+
+        [SerializeField]
+        [Tooltip("A fallback controller model to render in case the platform model fails to load")]
+        private GameObject fallbackControllerModel;
+
+        // caching the controller we belong to
+        private XRBaseController xrController;
+
+        // caching the controller model provider we are using we belong to
+        private ControllerModel controllerModelProvider;
+
+        // The controller usages we want the input device to have;
+        private InternedString targetUsage;
+
+        // The characteristics we want to check to ensure that the input device is for the appropriate hand.
+        private InputDeviceCharacteristics handednessCharacteristic;
+
+        // The controller characteristics we want the input device to have;
+        private InputDeviceCharacteristics validControllerCharacteristiscs = InputDeviceCharacteristics.HeldInHand & InputDeviceCharacteristics.TrackedDevice;
+
+        /// <inheritdoc />
+        protected bool IsControllerTracked => xrController.currentControllerState.inputTrackingState.HasPositionAndRotation();
+
+        protected void OnEnable()
```

Comment:
Any reason why these are `protected` but not `virtual`? Usually it'd be either both or neither, right?

Reply:
("these" meaning lifecycle methods)

Reply:
There's an interesting Unity code analyzer rule that recommends `protected` over `private` for inheritance reasons (derived classes inadvertently hiding child impls), but it looks like it's turned off by default: https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/UNT0021.md

We haven't really been consistent with this, but it's an interesting idea imo.